File: hw0/sandpile.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

void print_table(int table[23][23])
{
  int x, y;
  for (int y = 0; y < 23; y++)
  {
    for (int x = 0; x < 23; x++)
    {
      if (table[y][x] == -1)
      {
        printf("# ");
      }
      else
      {
        printf("%d ", table[y][x]);
      }
    }
    printf("\n");
  }
  printf("-----------------------------------------\n");
}

void initialize_table(int table[23][23], int argc, char *argv[])
{
  int i;
  if (argc > 1 && strcmp(argv[1], "--fps") == 0)
  {
    for (i = 3; i < argc; i += 3)
    {
      int y = atoi(argv[i]);
      int x = atoi(argv[i + 1]);
      int h = atoi(argv[i + 2]);
      table[y][x] = h;
    }
  }
  else
  {
    for (i = 1; i < argc; i += 3)
    {
      int y = atoi(argv[i]);
      int x = atoi(argv[i + 1]);
      int h = atoi(argv[i + 2]);
      table[y][x] = h;
    }
  }
}

void topple_cell(int table[23][23], int x, int y)
{
  int i;
  table[y][x] = 0;
  int neighbors[8][2] = {{y - 1, x - 1}, {y, x - 1}, {y + 1, x - 1}, {y - 1, x}, {y + 1, x}, {y - 1, x + 1}, {y, x + 1}, {y + 1, x + 1}};
  for (i = 0; i < 8; i++)
  {
    int u = neighbors[i][0];
    int v = neighbors[i][1];
    if (u < 0 || u > 21 || v < 0 || v > 21 || table[v][u] == -1)
    {
      continue;
    }
    if (table[v][u] >= 8)
    {
      topple_cell(table, u, v);
    }
    else
    {
      table[v][u] += 1;
    }
  }
}

void drop_sand_on_center(int table[23][23])
{
  if (table[11][11] >= 8)
  {
    topple_cell(table, 11, 11);
  }
  else
  {
    table[11][11] += 1;
  }
}

int main(int argc, char *argv[])
{
  int table[23][23] = {{0}};
  initialize_table(table, argc, argv);
  print_table(table);
  while (1)
  {
    drop_sand_on_center(table);
    print_table(table);
    if (argc > 1 && strcmp(argv[1], "--fps") == 0)
    {
      usleep(1000000 / atoi(argv[2]));
    }
    else
    {
      usleep(1000000);
    }
  }
}

// 30 fps = 0.5s sleep
// 60 fps = 1s sleep
// 10 fps = 1/6

File: hw0/sandpile.py
# take in all input sand piles, populate them into matrix
# iterate over entire matrix. for each cell, if < 8, += 1
# if == 8, add a value to each adjacent cell so long as it is not a sink
import time


class Sandpile:
    def __init__(self, inputs=[]):
        self.table = [[0 for i in range(22)] for j in range(22)]
        self.inputs = inputs
        self.initializeTable()
        self.printTable()
        while True:
            self.dropSandOnCenter()
            self.printTable()
            time.sleep(0.5)

    def initializeTable(self):
        for (x, y, h) in self.inputs:
            self.table[y][x] = h

    def dropSandOnCenter(self):
        if self.table[11][11] >= 8:
            self.toppleCell(11, 11)
        else:
            self.table[11][11] += 1

    def toppleCell(self, x, y):
        self.table[y][x] = 0
        neighbors = [(y-1, x-1), (y, x-1), (y+1, x-1), (y-1, x),
                     (y+1, x), (y-1, x+1), (y, x+1), (y+1, x+1)]
        for (u, v) in neighbors:
            if u < 0 or u > 21 or v < 0 or v > 21 or self.table[u][v] == -1:
                continue
            if self.table[v][u] >= 8:
                self.toppleCell(u, v)
            else:
                self.table[v][u] += 1

    def printTable(self):
        for y in range(22):
            line = ''
            for x in range(22):
                if self.table[y][x] == -1:
                    line += '#'
                else:
                    line += str(self.table[y][x])
            print(line)
        print('-----------------')


inputs = [(2, 2, -1), (14, 12, 5), (5, 16, 3), (12, 12, 6)]
sandpile = Sandpile(inputs)


